src/imp/security_framework.rs

```diff
@@ -38,6 +43,485 @@ lazy_static! {
+    CipherSuite::TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,
+    CipherSuite::TLS_DH_RSA_WITH_AES_256_GCM_SHA384,
+    CipherSuite::TLS_DHE_DSS_WITH_AES_256_GCM_SHA384,
+    CipherSuite::TLS_DH_DSS_WITH_AES_256_GCM_SHA384,
+    // CipherSuite::TLS_PSK_WITH_AES_256_GCM_SHA384,
+    // CipherSuite::TLS_DHE_PSK_WITH_AES_256_GCM_SHA384,
+    // CipherSuite::TLS_RSA_PSK_WITH_AES_256_GCM_SHA384,
+    // CipherSuite::TLS_PSK_WITH_AES_256_CBC_SHA384,
+    // CipherSuite::TLS_DHE_PSK_WITH_AES_256_CBC_SHA384,
+    // CipherSuite::TLS_RSA_PSK_WITH_AES_256_CBC_SHA384,
+    CipherSuite::TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384,
+    CipherSuite::TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384,
+    CipherSuite::TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384,
+    CipherSuite::TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384,
+    CipherSuite::TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+    CipherSuite::TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384,
+    CipherSuite::TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+    CipherSuite::TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384,
+];
+
+fn key_exchange_alg_to_cipher_suites(alg: &TlsKeyExchangeAlgorithm) -> &'static [CipherSuite] {
+    match alg {
+        TlsKeyExchangeAlgorithm::Dhe => CIPHERS_DH_EPHEM,
+        TlsKeyExchangeAlgorithm::Ecdhe => CIPHERS_ECDHE,
+        TlsKeyExchangeAlgorithm::Rsa => CIPHERS_RSA_KEYX,
+        TlsKeyExchangeAlgorithm::__NonExhaustive => unreachable!(),
+    }
+}
+
+fn signature_alg_to_cipher_suites(alg: &TlsSignatureAlgorithm) -> &'static [CipherSuite] {
+    match alg {
+        TlsSignatureAlgorithm::Dss => CIPHERS_DSS,
+        TlsSignatureAlgorithm::Ecdsa => CIPHERS_ECDSA,
+        TlsSignatureAlgorithm::Rsa => CIPHERS_RSA_SIGN,
+        TlsSignatureAlgorithm::__NonExhaustive => unreachable!(),
+    }
+}
+
+fn bulk_encryption_alg_to_cipher_suites(
+    alg: &TlsBulkEncryptionAlgorithm,
+) -> &'static [CipherSuite] {
+    match alg {
+        TlsBulkEncryptionAlgorithm::Aes128 => CIPHERS_AES_128,
+        TlsBulkEncryptionAlgorithm::Aes256 => CIPHERS_AES_256,
+        TlsBulkEncryptionAlgorithm::Des => CIPHERS_DES,
+        TlsBulkEncryptionAlgorithm::Rc2 => CIPHERS_RC2,
+        TlsBulkEncryptionAlgorithm::Rc4 => CIPHERS_RC4,
+        TlsBulkEncryptionAlgorithm::TripleDes => CIPHERS_3DES,
+        TlsBulkEncryptionAlgorithm::__NonExhaustive => unreachable!(),
+    }
+}
+
+fn hash_alg_to_cipher_suites(alg: &TlsHashAlgorithm) -> &'static [CipherSuite] {
+    match alg {
+        TlsHashAlgorithm::Md5 => CIPHERS_MD5,
+        TlsHashAlgorithm::Sha1 => CIPHERS_SHA1,
+        TlsHashAlgorithm::Sha256 => CIPHERS_SHA256,
+        TlsHashAlgorithm::Sha384 => CIPHERS_SHA384,
+        TlsHashAlgorithm::__NonExhaustive => unreachable!(),
+    }
+}
+
+fn expand_algorithms(cipher_suites: &CipherSuiteSet) -> Vec<CipherSuite> {
+    let first = cipher_suites
+        .key_exchange
+        .iter()
+        .flat_map(key_exchange_alg_to_cipher_suites)
+        .copied();
+    let rest: &[HashSet<_>] = &[
```

Comment:
It might be a bit better to load the key exchange ciphers into a hashset and then iterate through the the signature, bulk_encryption, etc ciphers to progressively remove bad ciphers from the set, and then load the remainder into the returned vec.

Reply:
That's a good point, I can change this.

Reply:
Made an attempt at this, although I'm not entirely convinced that it's better that it was before.